crates/arrow_sql_gen/src/clickhouse.rs

```diff
@@ -0,0 +1,546 @@
+    FailedToParseBigDecimalFromClickhouse {
+        value: String,
+        source: bigdecimal::ParseBigDecimalError,
+    },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+macro_rules! handle_primitive_type {
+    ($builder:expr, $type:expr, $builder_ty:ty, $value_ty:ty, $row:expr, $index:expr) => {{
+        let Some(builder) = $builder else {
+            return NoBuilderForIndexSnafu { index: $index }.fail();
+        };
+        let Some(builder) = builder.as_any_mut().downcast_mut::<$builder_ty>() else {
+            return FailedToDowncastBuilderSnafu {
+                clickhouse_type: format!("{:?}", $type),
+            }
+            .fail();
+        };
+        let v = $row
+            .get::<$value_ty, usize>($index)
+            .context(FailedToGetRowValueSnafu {
+                clickhouse_type: $type,
+            })?;
+
+        builder.append_value(v)
+    }};
+}
+
+macro_rules! handle_primitive_nullable_type {
+    ($builder:expr, $type:expr, $builder_ty:ty, $value_ty:ty, $row:expr, $index:expr) => {{
+        let Some(builder) = $builder else {
+            return NoBuilderForIndexSnafu { index: $index }.fail();
+        };
+        let Some(builder) = builder.as_any_mut().downcast_mut::<$builder_ty>() else {
+            return FailedToDowncastBuilderSnafu {
+                clickhouse_type: format!("{:?}", $type),
+            }
+            .fail();
+        };
+        let v = $row
+            .get::<Option<$value_ty>, usize>($index)
+            .context(FailedToGetRowValueSnafu {
+                clickhouse_type: $type,
+            })?;
+
+        match v {
+            Some(v) => builder.append_value(v),
+            None => builder.append_null(),
+        }
+    }};
+}
+
+/// Converts `Clickhouse` `Block` to an Arrow `RecordBatch`. Assumes that all rows have the same schema and
+/// sets the schema based on the `sql_type` returned for column.
+///
+/// # Errors
+///
+/// Returns an error if there is a failure in converting the rows to a `RecordBatch`.
+#[allow(clippy::too_many_lines)]
+pub fn block_to_arrow(block: &Block<Complex>) -> Result<RecordBatch> {
+    let mut arrow_fields: Vec<Option<Field>> = Vec::new();
+    let mut arrow_columns_builders: Vec<Option<Box<dyn ArrayBuilder>>> = Vec::new();
+    let mut clickhouse_types: Vec<SqlType> = Vec::new();
+    let mut column_names: Vec<String> = Vec::new();
+
+    if !block.is_empty() {
+        let columns = block.columns();
+        for column in columns {
+            let column_name = column.name();
+            let column_type = column.sql_type();
+            let data_type = map_column_to_data_type(&column_type);
+            arrow_fields.push(
+                data_type
+                    .clone()
+                    .map(|data_type| Field::new(column_name, data_type.clone(), true)),
+            );
+            arrow_columns_builders
+                .push(map_data_type_to_array_builder_optional(data_type.as_ref()));
```

Comment:
```suggestion
            let data_type = map_column_to_data_type(&column_type);
            arrow_fields.push(Some(Field::new(column_name, data_type.clone(), true)));
            arrow_columns_builders.push(map_data_type_to_array_builder_optional(Some(&data_type)));
```

After changing `map_column_to_data_type` to return a DataType directly instead of an Option<DataType>

Reply:
Fixed